compiler/rustc_span/src/hygiene.rs

```diff
@@ -572,35 +578,32 @@ impl HygieneData {
         }
 
         // Reserve a new syntax context.
-        // The inserted dummy data can only be potentially accessed by nested `alloc_ctxt` calls,
-        // the assert below ensures that it doesn't happen.
         let ctxt = SyntaxContext::from_usize(self.syntax_context_data.len());
         self.syntax_context_data
             .push(SyntaxContextData { dollar_crate_name: sym::dummy, ..SyntaxContextData::root() });
         self.syntax_context_map.insert(key, ctxt);
 
-        // Opaque and semi-opaque versions of the parent. Note that they may be equal to the
-        // parent itself. E.g. `parent_opaque` == `parent` if the expn chain contains only opaques,
-        // and `parent_opaque_and_semiopaque` == `parent` if the expn contains only (semi-)opaques.
+        // Get parent data
         let parent_data = &self.syntax_context_data[parent.0 as usize];
         assert_ne!(parent_data.dollar_crate_name, sym::dummy);
         let parent_opaque = parent_data.opaque;
         let parent_opaque_and_semiopaque = parent_data.opaque_and_semiopaque;
 
-        // Evaluate opaque and semi-opaque versions of the new syntax context.
+        // Evaluate opaque and semi-opaque versions iteratively to avoid recursive calls
         let (opaque, opaque_and_semiopaque) = match transparency {
             Transparency::Transparent => (parent_opaque, parent_opaque_and_semiopaque),
-            Transparency::SemiOpaque => (
-                parent_opaque,
-                // Will be the same as `ctxt` if the expn chain contains only (semi-)opaques.
-                self.alloc_ctxt(parent_opaque_and_semiopaque, expn_id, transparency),
-            ),
-            Transparency::Opaque => (
-                // Will be the same as `ctxt` if the expn chain contains only opaques.
-                self.alloc_ctxt(parent_opaque, expn_id, transparency),
-                // Will be the same as `ctxt` if the expn chain contains only (semi-)opaques.
-                self.alloc_ctxt(parent_opaque_and_semiopaque, expn_id, transparency),
-            ),
+            Transparency::SemiOpaque => {
+                let opaque = parent_opaque;
+                let semi_opaque =
```

Comment:
Could you keep the previous code structure (just tuples, no locals)?
The local names are incorrect (`opaque_and_semiopaque`, not `semi_opaque`) and do not improve readability here because the tuple parts are already named a few lines above.